gridcomps/History/MAPL_StationSamplerMod.F90

```diff
@@ -0,0 +1,461 @@
+!  Input Data:
+!__ 1. station_id_file: static
+!     plain text format: [id,name,lat,lon,elev]
+
+#include "MAPL_Generic.h"
+module StationSamplerMod
+  use ESMF
+  use MAPL_ErrorHandlingMod
+  use LocStreamFactoryMod
+  use pFIO
+  use MAPL_TimeDataMod
+  use MAPL_VerticalDataMod
+  use MAPL_BaseMod
+  use MAPL_CommsMod
+  use MAPL_LocstreamRegridderMod
+  use, intrinsic :: iso_fortran_env, only: REAL32
+  use, intrinsic :: iso_fortran_env, only: REAL64
+  implicit none
+  private
+
+  public :: StationSampler
+  type :: StationSampler
+     type(LocStreamFactory) :: LSF
+     type(ESMF_LocStream) :: esmf_ls 
+     type(LocstreamRegridder) :: regridder
+     type(ESMF_time), allocatable :: times(:)
+     integer, allocatable :: ids(:)
+     integer :: nobs
+     integer :: nstation
+     integer, allocatable :: station_id(:)
+     character(len=ESMF_MAXSTR), allocatable :: station_name(:)
+     real(kind=REAL64), allocatable :: lons(:)
+     real(kind=REAL64), allocatable :: lats(:)
+     real(kind=REAL64), allocatable :: elevs(:)
```

Comment:
Introduce a new derived type:
```f90
type Station
   integer :: id
   character(len=:), allocatable :: station_name
   real(kind=REAL64) :: lon
   real(kind=REAL64) :: lev
   real(kind=REAL64) :: elevation
end type
```

Reply:
We can introduce this OO type, but we still need the arrays lons(:) and lats(:)  because they are used later on 
*  creation of locStreamFactory as pass argument sampler%LSF = LocStreamFactory(sampler%lons, sampler%lats, _RC)
* pass to netCDF calls: call this%formatter%put_var('latitude',this%lats,_RC)

Reply:
First, this is not OO, just a derived type.   And if you have
```f90
type :: StationSampler
   ...
   type(Station), allocatable:: stations()
   ...
end type
...
call ... ('latitude', this%stations%lat, _RC)
```
it is the same as what you have.  Possibly a tad slower because of striding in memory, but that is not a concern here.

Reply:
I can give a try to see if this will work. I did not know this.

Reply:
@tclune,

I am trying exactly this path.  But the code failed to compile.  The subroutines were asking for explicit arrays, I think. 
May be the subroutine in F90/F77 is asking for the array head address then starting continuously from there. I am not sure. 

```
  type :: station
     integer :: station_id
     character(len=ESMF_MAXSTR), allocatable :: station_name
     real(kind=REAL64), allocatable :: lon
     real(kind=REAL64), allocatable :: lat
     real(kind=REAL64), allocatable :: elev
  end type station

  type :: StationSampler
     type(station), allocatable :: stations(:)
...

    allocate(sampler%stations(nstation))
    rewind(unit)
    do i=1, nstation
       read(unit, *) &
            sampler%stations(i)%station_id, &
            sampler%stations(i)%station_name, &
            sampler%stations(i)%lat, &
            sampler%stations(i)%lon

    sampler%LSF = LocStreamFactory(sampler%stations%lon, sampler%stations%lat, _RC)

    call this%formatter%put_var('longitude',this%stations%lon,_RC)
```

```
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)


  311 |     call this%formatter%put_var('latitude',this%stations%lat,_RC)
      |                                           1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)

  310 |     call this%formatter%put_var('longitude',this%stations%lon,_RC)
      |             1
Error: Found no matching specific binding for the call to the GENERIC ‘put_var’ at (1)

  105 |     sampler%LSF = LocStreamFactory(sampler%stations%lon, sampler%stations%lat, _RC)
      |                                   1
Error: Component to the right of a part reference with nonzero rank must not have the ALLOCATABLE attribute at (1)

```